core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -1281,13 +1389,323 @@ public String toStringResponseOnly(Function<? super ResponseHeaders, ?> headersS
     }
 
-    private static final class ListenerEntry {
-        final RequestLogListener listener;
+    private static final class RequestLogFuture extends EventLoopCheckingFuture<RequestLog> {
+
         final int interestedFlags;
 
-        ListenerEntry(RequestLogListener listener, int interestedFlags) {
-            this.listener = listener;
+        RequestLogFuture(int interestedFlags) {
             this.interestedFlags = interestedFlags;
         }
+
+        void completeLog(RequestLog log) {
+            super.complete(log);
+        }
+
+        @Override
+        public boolean complete(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public boolean cancel(boolean mayInterruptIfRunning) {
+            return false;
+        }
+
+        @Override
+        public void obtrudeValue(RequestLog value) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public void obtrudeException(Throwable ex) {
+            // Disallow users from completing arbitrarily.
+            throw new UnsupportedOperationException();
+        }
+    }
+
+    /**
+     * A {@link RequestLog} that delegates to {@link DefaultRequestLog} to read properties without
+     * checking availability.
+     */
+    private final class CompleteRequestLog implements RequestLog {
+
+        @Override
+        public boolean isComplete() {
+            return true;
+        }
+
+        @Override
+        public boolean isRequestComplete() {
+            return true;
+        }
+
+        @Override
+        public boolean isAvailable(RequestLogProperty property) {
+            return true;
+        }
+
+        @Override
+        public RequestLog partial() {
+            return this;
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> whenComplete() {
+            if (completedFuture == null) {
+                completedFuture = UnmodifiableFuture.completedFuture(this);
+            }
+            return completedFuture;
+        }
+
+        @Override
+        public CompletableFuture<RequestOnlyLog> whenRequestComplete() {
+            // OK to cast because the future is unmodifiable.
+            @SuppressWarnings("unchecked")
+            final CompletableFuture<RequestOnlyLog> cast =
+                    (CompletableFuture<RequestOnlyLog>) (CompletableFuture<?>) whenComplete();
+            return cast;
+        }
+
+        @Override
+        public CompletableFuture<RequestLog> whenAvailable(RequestLogProperty property) {
+            return whenComplete();
```

Comment:
ditto. What if a user does like:
```
log.whenAvailable(RequestLogProperty. REQUEST_FIRST_BYTES_TRANSFERRED_TIME).thenApply(log -> {this log might not have the property.})
```

Reply:
Again, `{REQUEST,RESPONSE}_FIRST_BYTES_TRANSFERRED_TIME` is not treated specially anymore, so the callback will be notified immediately.